sdk/search/Azure.Search.Documents/src/Batching/SearchIndexingBufferedSender{T}.cs

```diff
@@ -0,0 +1,606 @@
+                        // Call the service to find the name of the key
+                        SearchIndexClient indexClient = SearchClient.GetSearchIndexClient();
+                        SearchIndex index = async ?
+                            await indexClient.GetIndexAsync(IndexName, cancellationToken).ConfigureAwait(false) :
+                            indexClient.GetIndex(IndexName, cancellationToken);
+                        SearchField keyField = index.Fields.Single(f => f.IsKey == true);
+                        string key = keyField.Name;
+
+                        if (typeof(T).IsAssignableFrom(typeof(SearchDocument)))
+                        {
+                            // Case 3a: If it's a dynamic SearchDocument, lookup
+                            // the name of the key in the dictionary
+                            KeyFieldAccessor = (T doc) => (doc as SearchDocument)?.GetString(key);
+                            return;
+                        }
+                        else
+                        {
+                            // Case 3b: We'll see if there's a property with the
+                            // same name and use that as the accessor
+                            if (typeof(T).GetProperty(key) != null ||
+                                typeof(T).GetField(key) != null)
+                            {
+                                KeyFieldAccessor = CompileAccessor(key);
+                                return;
+                            }
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // We'll provide any exceptions as a hint because it could
+                        // be something like using the wrong API Key type when
+                        // moving from SearchClient up to SearchIndexClient that
+                        // potentially could be addressed if the user really wanted
+                        failure = ex;
+                    }
+
+                    // Case 4: Throw and tell the user to provide an explicit accessor.
+                    throw new InvalidOperationException(
+                        $"Failed to discover the Key field of document type {typeof(T).Name} for Azure Cognitive Search index {IndexName}.  " +
+                        $"Please set {typeof(SearchIndexingBufferedSenderOptions<T>).Name}.{nameof(SearchIndexingBufferedSenderOptions<T>.KeyFieldAccessor)} explicitly.",
+                        failure);
+                }
+                finally
+                {
+                    // Unblock anyone else who called this at the same time
+                    _keyFieldBarrier.Release();
+                }
+            }
+
+            // Build an accessor for a property named key on type T.  Compiling
+            // is kind of heavyweight, but we'll be calling this so many times
+            // it's worth it.
+            static Func<T, string> CompileAccessor(string key)
```

Comment:
This is rather heavyweight. Why not instead use the `PropertyInfo` or `FieldInfo` you used above to pass in the current document. It'd be a simple delegate instead of pulling in the assembly/ies for compiling expressions.

Reply:
We're going to call this on every single document.  It felt like one of those things that would amortize alright.  Plus I think I'd still have to wrap those rather than use the Getter directly.

Reply:
That's a great point.